File: typescript/packages/core/src/thread/control.ts
import type { z } from "zod";
import type { ParkAddress } from "../fold/state";
import {
  type BranchId,
  EventId,
  type JsonObject,
  type KnownEvent,
  type ModelRef,
  type Principal,
  principalKey,
} from "../log";
import { knownEvents } from "../reduce";
import { err, ok, type Result } from "../result";
import {
  type EventDraft,
  type LogStore,
  liveWriter,
  type Writer,
} from "../store";
import { uuidv7 } from "../store/encode";
import { isRefusal } from "../store/writer";
import { type Ask, invalidAnswer, matchAnswer } from "../tools/ask-user";
import type { Chain, ChainEvent, LogError } from "../verify";
import { askerOf } from "./questions";

// The Thread control methods (spec/api.json Thread): each appends the actor's
// event through the run's own writer when this process runs the branch, else under a short lease
// of its own; a lease another process holds is branch_busy, never waited on. The host
// authorizes the principal first (approver policy); here a principal of
// another tenant is refused, and the log's own rules decide the rest.

/** A control method's typed failure (its api.json errors). */
export type ControlError = {
  readonly code:
    | "forbidden"
    | "not_found"
    | "invalid_request"
    | "approval_mismatch"
    | "approval_expired"
    | "approval_duplicate"
    | "no_open_question"
    | "invalid_answer"
    | "not_parked"
    | "invalid_transition"
    | "branch_busy"
    | "branch_not_runnable";
  readonly message: string;
};

/** host-api Appended: the control operation's durable record. */
export type Appended = { readonly event_id: EventId };

/** host-api SettingsChange. */
export type SettingsChange = {
  readonly model: z.infer<typeof ModelRef>;
  readonly model_params?: z.infer<typeof JsonObject>;
  readonly reasoning_carryover?: "keep" | "omit_prior";
};

type Controlled = Result<Appended, ControlError>;
/** What one control appends: its record, then what follows it (a resumed naming the record). */
export type Plan = {
  readonly record: EventDraft;
  readonly after?: (record: EventId) => readonly EventDraft[];
};

const HOLDER = `control-${crypto.randomUUID()}`;

const fail = (code: ControlError["code"], message: string): Controlled =>
  err({ code, message });

export function resumed(address: ParkAddress, cause: EventId): EventDraft {
  return {
    type: "resumed",
    type_version: 1,
    critical: true,
    actor: { kind: "host" },
    data: { address, cause_event_id: cause },
  };
}

/** Host rows written in the same transaction as a control's record (a consumed inbox item). */
export type Alongside = Parameters<Writer["append"]>[1];

export type ControlOptions = {
  readonly alongside?: Alongside;
  /**
   * Only between turns (compact, setOutputStyle): a run in progress, one parked on an approval
   * or a question included, is branch_busy, and an inspection-only branch is branch_not_runnable.
   */
  readonly idle?: boolean;
};

const BUSY = "the thread has a turn in progress; try again when it ends";

/**
 * Appends the plan's events in one transaction, through the live run's writer or under a lease
 * taken and handed back here.
 */
export async function control(
  log: LogStore,
  branchId: BranchId,
  principal: Principal,
  plan: (
    events: readonly KnownEvent[],
    chain: Chain,
  ) => Result<Plan, ControlError>,
  { alongside, idle = false }: ControlOptions = {},
): Promise<Controlled> {
  if (principal.tenant !== log.tenant)
    return fail("forbidden", "the principal is not of this thread's tenant");
  const live = liveWriter(branchId);
  if (live !== undefined)
    return idle ? fail("branch_busy", BUSY) : appendPlan(live, plan, alongside);
  const writer = await log.acquire(branchId, HOLDER);
  if (!writer.ok) {
    // A run of this process took the lease while this acquire waited: through its writer.
    const started = liveWriter(branchId);
    if (started !== undefined && !idle)
      return appendPlan(started, plan, alongside);
    return fail(acquireCode(writer.error, idle), writer.error.message);
  }
  const planned = idle ? between(plan) : plan;
  try {
    return await appendPlan(writer.value, planned, alongside);
  } finally {
    await writer.value.release();
  }
}

/** The existing controls report a branch they can't run as not_found; the idle ones say why. */
function acquireCode(error: LogError, idle: boolean): ControlError["code"] {
  if (error.code === "branch_busy") return "branch_busy";
  return idle && error.code === "branch_not_runnable"
    ? "branch_not_runnable"
    : "not_found";
}

/** A parked run releases its lease, so the lease alone doesn't prove the thread is idle. */
function between(
  plan: (
    events: readonly KnownEvent[],
    chain: Chain,
  ) => Result<Plan, ControlError>,
): (events: readonly KnownEvent[], chain: Chain) => Result<Plan, ControlError> {
  return (events, chain) =>
    chain.fold.turnOpen
      ? err({ code: "branch_busy", message: BUSY })
      : plan(events, chain);
}

/**
 * The plan, decided under the writer's lock against the chain it extends (a control queued
 * behind a live run's in-flight append plans against what that append left), then appended.
 */
async function appendPlan(
  writer: Writer,
  plan: (
    events: readonly KnownEvent[],
    chain: Chain,
  ) => Result<Plan, ControlError>,
  alongside?: Alongside,
): Promise<Controlled> {
  const done = await writer.appendDecided(async ({ chain, now }) => {
    const planned = plan(knownEvents(chain), chain);
    if (!planned.ok) return planned;
    const { record, after } = planned.value;
    // One append: the record names its own id, so what follows it can cite it.
    const id = EventId.parse(uuidv7(now));
    return ok([{ ...record, event_id: id }, ...(after?.(id) ?? [])]);
  }, alongside);
  if (isRefusal(done)) return err(done.refusal);
  if (!done.ok) return fail(codeOf(done.error), done.error.message);
  return ok({ event_id: eventIdOf(done.value[0]) });
}

function eventIdOf(line: ChainEvent | undefined): EventId {
  if (line?.kind !== "event") throw new Error("a control appends known events");
  return line.event.event_id;
}

function codeOf(error: LogError): ControlError["code"] {
  switch (error.code) {
    case "approval_mismatch":
    case "approval_expired":
    case "approval_duplicate":
    case "invalid_transition":
    case "branch_busy":
      return error.code;
    default:
      return "invalid_request";
  }
}

/** A resumed for `address` when the branch is parked on it. */
export function resumeIf(
  chain: Chain,
  address: ParkAddress,
): ((id: EventId) => readonly EventDraft[]) | undefined {
  const parked = chain.fold.parked.some(
    (a) => a.kind === address.kind && a.id === address.id,
  );
  return parked ? (id) => [resumed(address, id)] : undefined;
}

/** Whether `principal` gave the user_input that opened the turn with the question's call. */
function asker(
  events: readonly KnownEvent[],
  callId: string,
  principal: Principal,
): boolean {
  const opener = askerOf(events, callId);
  return (
    opener !== undefined && principalKey(opener) === principalKey(principal)
  );
}

const FREE: Ask = { question: "" };

/** answer: the ask_user call's result, from the asking principal, strict to its options. */
export function answer(
  callId: string,
  text: string | readonly string[],
  principal: Principal,
): (events: readonly KnownEvent[], chain: Chain) => Result<Plan, ControlError> {
  return (events, chain) => {
    const address: ParkAddress = { kind: "input", id: callId };
    const after = resumeIf(chain, address);
    if (after === undefined)
      return err({
        code: "no_open_question",
        message: `no question ${callId}`,
      });
    if (!asker(events, callId, principal))
      return err({
        code: "forbidden",
        message: "only the user whose input opened this turn may answer",
      });
    const asked = chain.fold.asks.get(callId);
    // Rule 48 keeps an unreadable question from parking; one that did anyway takes free text.
    const ask = asked === undefined || asked === "invalid" ? FREE : asked;
    const recorded = matchAnswer(ask, text);
    if (recorded === undefined)
      return err({ code: "invalid_answer", message: invalidAnswer(ask) });
    return ok({
      record: {
        type: "tool_result",
        type_version: 1,
        critical: true,
        actor: { kind: "user", principal },
        data: {
          call_id: callId,
          is_error: false,
          origin: "answered",
          completeness: "complete",
          preview: recorded,
        },
      },
      after,
    });
  };
}

/** resolveParked: a human settles a parked effect (C3). */
export function resolveParked(
  effectKey: string,
  resolution: "assume_done" | "assume_not_done",
  principal: Principal,
): (events: readonly KnownEvent[], chain: Chain) => Result<Plan, ControlError> {
  return (_events, chain) => {
    const address: ParkAddress = { kind: "effect", id: effectKey };
    const after = resumeIf(chain, address);
    const effect = chain.fold.effects.get(effectKey);
    if (after === undefined || effect === undefined)
      return err({ code: "not_parked", message: `${effectKey} is not parked` });
    return ok({
      record: {
        type: "effect_resolved",
        type_version: 1,
        critical: true,
        actor: { kind: "approver", principal },
        data: { call_id: effect.callId, outcome: resolution, by: "human" },
      },
      after,
    });
  };
}

File: python/src/threads/thread/control.py
"""Thread controls (spec/api.json `Thread`: answer, resolve_parked, cancel, set_model, set_mode):
each appends one durable record under the branch's single writer, with the acting principal as
its actor. The record is the control; a run picks it up from the log.

A run in flight in this process owns the branch's lease, so a control appends through that
run's writer (`LIVE`); otherwise it takes the lease for the one append and hands it back.
"""

import uuid
from collections.abc import Callable, Sequence
from typing import TYPE_CHECKING, Final, Literal

from pydantic.experimental.missing_sentinel import MISSING

from threads._generated.host_api_v1 import Appended, SettingsChange
from threads._generated.tools_v1 import AskUserInput
from threads.agents.store import LIVE, Store, now_ms, open_store
from threads.log import (
    BranchId,
    CallId,
    ModelSettings,
    ParkAddress,
    ParseError,
    PermissionMode,
    Principal,
    SettingsChangedEvent,
    ThreadStartedEvent,
    UserInputEvent,
)
from threads.log.ask_user import Ask, ask_of, invalid_answer, match_answer
from threads.loop.runtime import LOST
from threads.reduce import Fold
from threads.reduce.handlers import to_json
from threads.result import Err, Ok
from threads.store import Draft, StoredEvent, Writer
from threads.store.companion import Companion
from threads.store.lines import uuid7

if TYPE_CHECKING:
    from pydantic import JsonValue

LOCAL_OPERATOR: Final = Principal(issuer="api", tenant="local", subject="operator")
"""The default principal of a local run, and its default approver."""

type Controlled = Ok[Appended] | Err[ParseError]
type Build = Callable[[Fold], Ok[Sequence[Draft]] | Err[ParseError]]


def actor(kind: Literal["user", "approver"], principal: Principal) -> "dict[str, JsonValue]":
    return {"kind": kind, "principal": to_json(principal)}


def forbidden(why: str) -> Err[ParseError]:
    return Err(ParseError("forbidden", why))


BUSY: Final = "the thread has a turn in progress; try again when it ends"


async def append(
    store: Store,
    branch: BranchId,
    build: Build,
    companion: Companion | None = None,
    *,
    idle: bool = False,
) -> Controlled:
    """Builds the drafts from the committed branch and appends them as one transaction. The
    first draft is the control's durable record. `idle`: only between turns (a turn parked on an
    approval, a question or an effect is still open), so never through a live run's writer."""

    async def under(writer: Writer) -> Controlled:
        if idle and writer.fold.in_turn:
            return Err(ParseError("branch_busy", BUSY))
        drafts = build(writer.fold)
        if isinstance(drafts, Err):
            return drafts
        done = await writer.append(drafts.value, companion)
        if isinstance(done, Err):
            if done.error.code in LOST:
                # The run here lost the branch to another holder: to the caller that is the
                # same as a lease held elsewhere.
                return Err(ParseError("branch_busy", done.error.message))
            return done
        return Ok(Appended(event_id=done.value[0].event_id))

    live = LIVE.get(branch)
    if live is not None:
        return Err(ParseError("branch_busy", BUSY)) if idle else await under(live)
    sq = await open_store(store)
    taken = await sq.acquire(branch, uuid.uuid4().hex, now_ms)
    if isinstance(taken, Err):
        return taken
    try:
        return await under(taken.value)
    finally:
        await taken.value.release()


def resumed(fold: Fold, address: ParkAddress, cause: str) -> tuple[Draft, ...]:
    """A resumed for the address when the branch is parked on it."""
    if address not in fold.parked:
        return ()
    data: dict[str, JsonValue] = {"address": to_json(address), "cause_event_id": cause}
    return (Draft("resumed", data),)


def first(kind: str, data: "dict[str, JsonValue]", by: "dict[str, JsonValue]") -> Draft:
    return Draft(kind, data, by, True, uuid7(now_ms()))


async def cancel(
    store: Store,
    branch: BranchId,
    principal: Principal,
    *,
    kind: Literal["cancel_requested", "stop_when_idle"] = "cancel_requested",
    companion: Companion | None = None,
) -> Controlled:
    """A durable cancel_requested: the run stops at its next step, and an effect in doubt is
    settled or parked first, never cancelled over. stop_when_idle (a channel's soft stop)
    finishes what is in flight and starts nothing new."""
    if principal.tenant != store.tenant:
        return forbidden("another tenant's thread")
    if kind == "stop_when_idle":
        data: dict[str, JsonValue] = {"reason": "requested by the thread's principal"}
        soft = Draft(kind, data, actor("user", principal))
        return await append(store, branch, lambda _: Ok((soft,)), companion)
    barrier = first(kind, {"scope": "thread"}, actor("user", principal))
    return await append(store, branch, lambda fold: Ok(barred(fold, barrier)), companion)


def barred(fold: Fold, barrier: Draft) -> tuple[Draft, ...]:
    """The cancel_requested barrier, then a resumed for each open question or approval: those
    can't be answered after the barrier, so the loop closes their calls. An effect in doubt
    stays parked."""
    waiting = [a for a in fold.parked if a.kind in ("approval", "input")]
    return (barrier, *(d for a in waiting for d in resumed(fold, a, _id(barrier))))


async def set_mode(
    store: Store, branch: BranchId, mode: PermissionMode, principal: Principal
) -> Controlled:
    """mode_changed from the current mode; bypass needs allow_bypass, which
    the log's own rules check (invalid_transition)."""
    if principal.tenant != store.tenant:
        return forbidden("another tenant's thread")

    def build(fold: Fold) -> Ok[Sequence[Draft]]:
        data: dict[str, JsonValue] = {"from": fold.mode, "to": mode}
        return Ok((Draft("mode_changed", data, actor("user", principal)),))

    return await append(store, branch, build)


async def set_model(
    store: Store, branch: BranchId, settings: SettingsChange, principal: Principal
) -> Controlled:
    """settings_changed{reason: user}: a new settings epoch. The adapter and,
    by default, the model params carry over from the current epoch."""
    if principal.tenant != store.tenant:
        return forbidden("another tenant's thread")

    def build(fold: Fold) -> Ok[Sequence[Draft]] | Err[ParseError]:
        current = _settings(fold.events)
        if current is None:
            return Err(ParseError("invalid_transition", "the thread has not started"))
        params = settings.model_params
        carryover = settings.reasoning_carryover
        new = ModelSettings(
            model=settings.model,
            model_params=current.model_params if params is MISSING else params,
            adapter=current.adapter,
            reasoning_carryover="keep" if carryover is MISSING else carryover,
        )
        data: dict[str, JsonValue] = {"reason": "user", "settings": to_json(new)}
        return Ok((Draft("settings_changed", data, actor("user", principal)),))

    return await append(store, branch, build)


def _settings(events: Sequence[StoredEvent]) -> ModelSettings | None:
    for event in reversed(events):
        if isinstance(event, SettingsChangedEvent):
            return event.data.settings
        if isinstance(event, ThreadStartedEvent):
            data = event.data
            return ModelSettings(
                model=data.model,
                model_params=data.model_params,
                adapter=data.adapter,
                reasoning_carryover="keep",
            )
    return None


def requester(events: Sequence[StoredEvent]) -> Principal | None:
    """The principal whose input opened the current turn."""
    for event in reversed(events):
        if isinstance(event, UserInputEvent):
            return event.actor.principal
    return None


def question(fold: Fold, call_id: str) -> Ask:
    """The open question's ask_user input. Rule 46 keeps an unreadable one from parking; one that
    parked anyway takes free text."""
    call = fold.calls.get(CallId(call_id))
    ask = None if call is None else ask_of(dict(call.data.input))
    return FREE if ask is None else ask


FREE: Final = AskUserInput(question="?")


def answering(
    fold: Fold, call_id: str, text: str | Sequence[str], principal: Principal
) -> Ok[str] | Err[ParseError]:
    """The recorded answer to an open question from `principal`, strict to its options."""
    if ParkAddress(kind="input", id=call_id) not in fold.parked:
        return Err(ParseError("no_open_question", f"no open question {call_id}"))
    if requester(fold.events) != principal:
        return forbidden("only the principal who asked may answer")
    ask = question(fold, call_id)
    recorded = match_answer(ask, text)
    if recorded is None:
        return Err(ParseError("invalid_answer", invalid_answer(ask)))
    return Ok(recorded)


def answered(fold: Fold, call_id: str, preview: str, principal: Principal) -> tuple[Draft, ...]:
    """tool_result{origin: answered} by the asker, then resumed."""
    data: dict[str, JsonValue] = {
        "call_id": call_id,
        "is_error": False,
        "completeness": "complete",
        "preview": preview,
        "origin": "answered",
    }
    result = first("tool_result", data, actor("user", principal))
    address = ParkAddress(kind="input", id=call_id)
    return (result, *resumed(fold, address, _id(result)))


async def answer(
    store: Store,
    branch: BranchId,
    call_id: str,
    text: str | Sequence[str],
    principal: Principal,
) -> Controlled:
    """The answer to an open ask_user question: tool_result{origin: answered} by the principal
    whose input opened the turn, then resumed. With options only an option is an answer."""
    if principal.tenant != store.tenant:
        return forbidden("another tenant's thread")

    def build(fold: Fold) -> Ok[Sequence[Draft]] | Err[ParseError]:
        recorded = answering(fold, call_id, text, principal)
        if isinstance(recorded, Err):
            return recorded
        return Ok(answered(fold, call_id, recorded.value, principal))

    return await append(store, branch, build)


def _id(draft: Draft) -> str:
    if draft.event_id is None:
        raise AssertionError("a cause is appended with a preset id")
    return draft.event_id


async def resolve_parked(
    store: Store,
    branch: BranchId,
    effect_key: str,
    resolution: Literal["assume_done", "assume_not_done"],
    principal: Principal,
) -> Controlled:
    """A human settles an effect in doubt: effect_resolved{by: human}, then resumed. The caller has
    checked the principal's authority (threads.thread.authority)."""
    address = ParkAddress(kind="effect", id=effect_key)

    def build(fold: Fold) -> Ok[Sequence[Draft]] | Err[ParseError]:
        if address not in fold.parked:
            return Err(ParseError("not_parked", f"no parked effect {effect_key}"))
        call = effect_key.split(":", 1)[-1]
        data: dict[str, JsonValue] = {"call_id": call, "outcome": resolution, "by": "human"}
        settled = first("effect_resolved", data, actor("approver", principal))
        return Ok((settled, *resumed(fold, address, _id(settled))))

    return await append(store, branch, build)
